File: C++/16120_PPAP.cpp
#include <iostream>
#include <cstring>
#include <vector>
using namespace std;

int main() {
	string s;
	cin >> s;
	
	vector<char> v; // stack 대신 사용할 vector
	char PPAP[4] = {'P','A','P','P'}; // 뒤부터 PPAP가 맞는지 확인하기 위해 거꾸로 넣어 놓기
	
	int s_len = s.size();
	for(int i = 0; i < s_len; i++){
		if(s[i] == 'P'){ // P라면 그냥 삽입
			v.push_back(s[i]);
			continue;
		}
		
		// A라면 PPAP 문자열이 만들어지는 지 확인
		bool flag = true; // PPAP 문자가 만들어지는지 여부를 나타내는 flag
		
		v.push_back(s[i++]); // A 삽입
        if(i >= s_len){ // 범위 초과 여부 예외처리
            break;
        }
		v.push_back(s[i]); // A 다음 문자열 삽입(PPAP가 되려면 A 다음에 P가 삽입되기를 기대)
		for(int j = 0; j < 4; j++){ // PPAP가 되는지 확인
			if(PPAP[j] == v.back()){
				v.pop_back(); // 확인해서 일치하면 pop
			}
			else{ // 일치하지 않으면 PPAP 문자열이 아님 -> 더 이상 볼 필요 없음
				flag = false;
				break;
			}
		}
		if(!flag){
			break;
		}
		
		v.push_back('P'); // PPAP가 만들어졌으면 PPAP를 모두 빼고 P 넣기
	}
	
	if(v.size() == 1 && v[0] == 'P'){ // stack이 P 1개로만 이루어져 있다면
		cout << "PPAP"; // PPAP 문자열
	}
	else{ // 아니라면
		cout << "NP"; // PPAP 문자열이 아님
	}
	
	return 0;
}

File: Python/16120_PPAP.py
import sys

if __name__ == '__main__':
	PPAP = input()
	
	stack = []
	for i in PPAP:
		stack.append(i) # 일단 문자열 추가
		
		# PPAP인지 확인해야 하므로, 문자열이 4개 이상일 때부터 확인
		if len(stack) >= 4 and stack[-4:] == ['P', 'P', 'A', 'P']: # 가장 끝에 있는 문자열 4개가 PPAP라면
			for _ in range(3):
				stack.pop() # P 하나 남기고 모두 pop(4개 pop하고 P 한 개 넣는 것과 같음)
		
	print("PPAP" if len(stack) == 1 and stack[0] == 'P' else "NP") # stack에 P 하나만 남아있으면 PPAP 문자열, 아니라면 PPAP 문자열이 아니므로 NP 출력
